extensions/pkg/controller/bastion/reconciler.go

```diff
@@ -0,0 +1,154 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+
+	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/controllerutils"
+)
+
+type reconciler struct {
+	logger   logr.Logger
+	actuator Actuator
+
+	client        client.Client
+	reader        client.Reader
+	statusUpdater extensionscontroller.StatusUpdater
+}
+
+// NewReconciler creates a new reconcile.Reconciler that reconciles
+// bastion resources of Gardener's `extensions.gardener.cloud` API group.
+func NewReconciler(actuator Actuator) reconcile.Reconciler {
+	logger := log.Log.WithName(ControllerName)
+
+	return extensionscontroller.OperationAnnotationWrapper(
+		func() client.Object { return &extensionsv1alpha1.Bastion{} },
+		&reconciler{
+			logger:        log.Log.WithName(ControllerName),
+			actuator:      actuator,
+			statusUpdater: extensionscontroller.NewStatusUpdater(logger),
+		},
+	)
+}
+
+func (r *reconciler) InjectFunc(f inject.Func) error {
+	return f(r.actuator)
+}
+
+func (r *reconciler) InjectClient(client client.Client) error {
+	r.client = client
+	r.statusUpdater.InjectClient(client)
+	return nil
+}
+
+func (r *reconciler) InjectAPIReader(reader client.Reader) error {
+	r.reader = reader
+	return nil
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	bastion := &extensionsv1alpha1.Bastion{}
+	if err := r.client.Get(ctx, request.NamespacedName, bastion); err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	cluster, err := extensionscontroller.GetCluster(ctx, r.client, bastion.Namespace)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if extensionscontroller.IsFailed(cluster) {
+		r.logger.Info("Stop reconciling Bastion of failed Shoot.", "name", bastion.Name)
+		return reconcile.Result{}, nil
+	}
```

Comment:
I guess this is not needed for the same reason why the `LastOperationFailed` predicate was removed - the `Bastion` is not "tight" to the shoot cluster reconciliation, as opposed to most other resources.